### PR TITLE
WIP: galera support

### PR DESCRIPTION
**Pull request by jsuchome**
Experimental PR based on @rhafer 's work

### CODE REVIEW

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+resource_agent = "ocf:heartbeat:galera"
+
+package "galera-3"
```
Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: Does not seem to be needed with mariadb 10.1 (that provides galera plugin itself)

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+end
+
+service_name = "galera"
+
+pacemaker_primitive service_name do
+  agent resource_agent
+  params (
+    {
+      "enable_creation" => true,
+      "wsrep_cluster_address" => "gcomm://d52-54-00-01-00-01,d52-54-00-01-00-02,d52-54-00-01-00-03",
```
Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: TODO: generate from cluster membets

File: chef/cookbooks/mysql/templates/default/my.cnf.erb [hunk 3]
```diff
@@ -52,6 +52,16 @@ datadir                = <%= node['mysql']['datadir'] %>
 tmpdir                 = <%= node['mysql']['tmpdir'] %>
 skip-external-locking
 
+# GALERA
+
+wsrep_provider=/usr/lib64/galera-3/libgalera_smm.so
```
Comment (jsuchome) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: Seems like this could be be dropped with mariadb-10.1 (and points to old library anyway)

File: chef/cookbooks/mysql/templates/default/my.cnf.erb [hunk 4]
```diff
@@ -52,6 +52,16 @@ datadir                = <%= node['mysql']['datadir'] %>
...
+wsrep_cluster_address="gcomm://d52-54-00-01-00-01,d52-54-00-01-00-02,d52-54-00-01-00-03"
```
Comment (jsuchome) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: TODO: get via argument

File: crowbar_framework/app/models/database_service.rb [hunk 5]
```diff
@@ -73,33 +73,35 @@ def validate_proposal_after_save(proposal)
...
-        validation_error I18n.t(
-          "barclamp.#{@bc_name}.validation.drbd_not_enabled",
-          cluster_name: cluster_name(cluster)
-        ) unless role.default_attributes["pacemaker"]["drbd"]["enabled"]
+      unless db_engine == "mysql"
+        storage_mode = attributes["ha"]["storage"]["mode"]
         validation_error I18n.t(
-          "barclamp.#{@bc_name}.validation.invalid_size_drbd"
-        ) if attributes["ha"]["storage"]["drbd"]["size"] <= 0
+                           "barclamp.#{@bc_name}.validation.unknown_mode_ha",
```
Comment (s-t-e-v-e-n-k) [inline crowbar_framework/app/models/database_service.rb]: I think this whole block needs a rethink, and could probably do with some cleanup and reflowing. Secondly, I also think we want to check if we have <3 cluster members and error if sql_engine is mysql.

Comment (jsuchome) [inline crowbar_framework/app/models/database_service.rb]: I agree, but it might be enough for some next PR

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    subscribes :create,
+      "file[delete mariadb.service]",
+      :immediately
+    action :nothing
+  end
+
+  file "delete mariadb.service" do
+    path "/etc/systemd/system/mariadb.service"
+    action :delete
+    only_if
```
Comment (Itxaka) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: empty only_if

File: chef/cookbooks/mysql/recipes/ha_galera.rb (hunk 2 above)
Comment (Itxaka) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: hardcoded addresses?

File: chef/cookbooks/mysql/templates/default/my.cnf.erb (hunk 4 above)
Comment (Itxaka) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: harcoded addresses?

File: chef/cookbooks/mysql/recipes/ha_galera.rb (hunk 2 above)
Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: yep, see my older comment

File: crowbar_framework/app/models/database_service.rb [hunk 7]
```diff
@@ -60,6 +60,53 @@ def create_proposal
     base
   end
 
+  def validate_ha_attributes(attributes)
```
Comment (jsuchome) [inline crowbar_framework/app/models/database_service.rb]: I'm afraid I can't do anything about it. Splitting the method even more would not bring clarity IMHO.

File: crowbar_framework/app/models/database_service.rb [hunk 8]
```diff
@@ -60,6 +60,48 @@ def create_proposal
...
+            cluster_name: cluster_name(cluster)
+          )
+        end
+        if attributes["ha"]["storage"]["drbd"]["size"] <= 0
+          validation_error I18n.t(
+            "barclamp.#{@bc_name}.validation.invalid_size_drbd"
+          )
+        end
+      end
+    when "mysql"
```
Comment (jsuchome) [inline crowbar_framework/app/models/database_service.rb]: No more empty body

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+# all the required packages installed before we create the pacemaker
+# resources
+crowbar_pacemaker_sync_mark "sync-database_before_ha" do
+  revision node[:database]["crowbar-revision"]
+end
+
+unless node[:database][:galera_bootstrapped]
+  mysql_service_name = "mariadb"
+
+  bash "reload systemd after mariadb.service  update" do
```
Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: Nit: Double space after service

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  end
+
+  directory "/var/run/mysql/" do
+    owner "mysql"
+    group "root"
+    mode "0755"
+    action :create
+  end
+
+  # Temporary install mariadb unit as the "galera_new_cluster" bootstrap
```
Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: We could fix this by adding Alias=mariadb.service into the service file we ship in the mariadb package, making this block unnecessary.

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  file "mariadb.service" do
+    path "/etc/systemd/system/mariadb.service"
+    content ::File.open("/usr/share/mysql/systemd/mariadb.service").read
+    mode "0644"
+  end
+
+  execute "mysql_install_db" do
+    command "mysql_install_db"
+    action :run
+    # only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
```
Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: So, founder or all cluster nodes?

File: chef/cookbooks/mysql/recipes/ha_galera.rb [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+end
+
+transaction_objects.push("pacemaker_primitive[#{service_name}]")
+
+ms_name = "ms-#{service_name}"
+pacemaker_ms ms_name do
+  rsc service_name
+  meta(
+    "master-max" => 3,
```
Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: This should be the number of cluster members, not hardcoded, IMO.

File: chef/cookbooks/mysql/templates/default/my.cnf.erb [hunk 13]
```diff
@@ -52,6 +52,15 @@ datadir                = <%= node['mysql']['datadir'] %>
 tmpdir                 = <%= node['mysql']['tmpdir'] %>
 skip-external-locking
 
+# GALERA
+
+wsrep_cluster_address="gcomm://d52-54-00-01-00-01,d52-54-00-01-00-02,d52-54-00-01-00-03"
```
Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: This is exactly same format as the wsrep_cluster_address in the pacemaker primitive. This should be filled in by the recipe.

File: crowbar_framework/config/locales/database/en.yml [hunk 14]
```diff
@@ -45,9 +45,9 @@ en:
               options: 'Mount Options'
       validation:
         invalid_db_engine: 'Invalid database engine: %{db_engine}.'
-        ha_postgresql: 'High availability support is only available for PostgreSQL.'
         unknown_mode_ha: 'Unknown mode for HA storage: %{storage_mode}.'
         no_device: 'No device specified for shared storage.'
         no_filesystem: 'No filesystem type specified for shared storage.'
         drbd_not_enabled: 'DRBD is not enabled for cluster %{cluster_name}.'
         invalid_size_drbd: 'Invalid size for DRBD device.'
+        cluster_size_even: 'The Galera cluster needs odd number of nodes.'
```
Comment (s-t-e-v-e-n-k) [inline crowbar_framework/config/locales/database/en.yml]: "needs an odd number of cluster members"

File: chef/cookbooks/mysql/recipes/ha_galera.rb (hunk 10 above)
Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: Or maybe there's some packaging bug?
If _"galera_new_cluster" bootstrap script doesn't work with installed "mysql.service" unit_ , that does not look correct...

File: chef/cookbooks/mysql/recipes/ha_galera.rb (hunk 12 above)
Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: Ah, good point

File: chef/cookbooks/mysql/templates/default/my.cnf.erb (hunk 13 above)
Comment (jsuchome) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: Yep

File: crowbar_framework/config/locales/database/en.yml (hunk 14 above)
Comment (jsuchome) [inline crowbar_framework/config/locales/database/en.yml]: OK, I'll update it

**Comment by jsuchome** [comment]
@s-t-e-v-e-n-k I was finally able to deploy database barclamp with mysql and HA. It suffers from the problem already mentioned by @rhafer (problem with promoting the nodes). At least this is something you could work with I hope

File: chef/cookbooks/mysql/templates/default/my.cnf.erb [hunk 15]
```diff
@@ -52,6 +52,16 @@ datadir                = <%= node['mysql']['datadir'] %>
...
 
+<% if node[:database][:ha][:enabled] %>
+# GALERA
+
+wsrep_cluster_address="gcomm://"<%= @nodes_names.join(",") %>
+innodb_autoinc_lock_mode=2
+innodb_doublewrite=1
+query_cache_size=0
+wsrep_on=ON
+binlog_format=ROW
```
Comment (rsalevsky) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: we should add `wsrep_sst_method=xtrabackup`

Comment (jsuchome) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: Why this and not anything else (and not the default)?

File: chef/cookbooks/mysql/recipes/ha_galera.rb (hunk 10 above)
Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: This is not required at all. From galera_new_cluster it uses systemctl start ${1:-mariadb} so we should call 'galera_new_cluster mysql' and all this temporary stuff can hit /dev/null.

Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: I think you are wrong. I did not check this closely, but when I really removed it, mysql was not started.... Oh, it was probably on the other (non-founder) nodes, that need to start mysql during bootstrap as well.

Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: Really? The service file is removed after bootstrapping. I've been testing a diff that removes all these gubbins.

Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: But the database needs to be started on all nodes during bootstrapping, to set the root password

Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: The only reason this service hackery exists is for galera_new_cluster, which we only run on the cluster founder, and if we call it with 'mysql' as the first argument (as opposed to no arguments), it will use that service name. The second piece of the puzzle is updating mysql_service_name to be "mysql" and it works great.

Comment (jsuchome) [inline chef/cookbooks/mysql/recipes/ha_galera.rb]: > The second piece of the puzzle is updating mysql_service_name to be "mysql" and it works great.

Ah, true. That could work.

File: chef/cookbooks/mysql/templates/default/my.cnf.erb [hunk 16]
```diff
@@ -52,6 +52,17 @@ datadir                = <%= node['mysql']['datadir'] %>
 tmpdir                 = <%= node['mysql']['tmpdir'] %>
 skip-external-locking
 
+<% if node[:database][:ha][:enabled] %>
+# GALERA
+
+wsrep_provider=/usr/lib64/galera-3/libgalera_smm.so
```
Comment (s-t-e-v-e-n-k) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: This file doesn't exist without the galera-3 package installed, so should be removed at the same time as the package "galera-3" statement in the recipe.

Comment (jsuchome) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: I've brought "galera-3" package back (see that the commit got removed). I was unable to finish the configuration without providing `wsrep_provider`

**Comment by jsuchome** [comment]
@s-t-e-v-e-n-k not that I removed again that commit that was removing that temporary service. I probably did something wrong there, but it was breaking my deployment.

File: chef/cookbooks/mysql/templates/default/my.cnf.erb (hunk 15 above)
Comment (rsalevsky) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: It's the only non-blocking SST method; see http://galeracluster.com/documentation-webpages/sst.html

Comment (jsuchome) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: @rsalevsky  Hm. I did not test too long, but with any non-default value of `wsrep_sst_method` I was not able to start the galera cluster.

Comment (rsalevsky) [inline chef/cookbooks/mysql/templates/default/my.cnf.erb]: Yeah, I tend to say we should do it in a separate PR. Otherwise this PR can get very messy.

**Comment by jsuchome** [comment]
Closing, was moved to https://github.com/crowbar/crowbar-openstack/pull/1111
